### PR TITLE
Modify flow analysis so that there is an implicit "break" after every switch section.

### LINKED ISSUE #612: Cascaded unreachable code diagnostic after missing break in case block
Originally from http://roslyn.codeplex.com/workitem/330

----------------------------------------
SSL reported:
```C#
switch (1) {
case 1:
}
int i;
```
The `case` label correctly gives an error for the missing `break`.
However, the `int i;` statement then gives an incorrect Unreachable Code warning.

### PR DESCRIPTION
**Pull request by gafter**
Because the end of a switch section is required not to be reachable by the
language spec, this only affects the results in error situations by suppressing
cascaded diagnostics. Fixes #612.

### CODE REVIEW

**Comment by gafter** [comment]
@jaredpar @VSadov @AlekseyTs @agocke @VladimirReshetnikov Could a couple of you review this please?

**Comment by VladimirReshetnikov** [comment]
:+1:

File: src/Compilers/CSharp/Portable/FlowAnalysis/PreciseAbstractFlowPass.cs [hunk 1]
```diff
@@ -1661,13 +1662,17 @@ private LocalState VisitSwitchHeader(BoundSwitchStatement node)
...
         private void VisitSwitchBlock(BoundSwitchStatement node)
         {
+            var afterSwitchState = UnreachableState();
             var switchSections = node.SwitchSections;
             var iLastSection = (switchSections.Length - 1);
             // visit switch sections
             for (var iSection = 0; iSection <= iLastSection; iSection++)
             {
                 VisitSwitchSection(switchSections[iSection], iSection == iLastSection);
+                IntersectWith(ref afterSwitchState, ref this.State);
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/FlowAnalysis/PreciseAbstractFlowPass.cs]: Consider adding a comment explaining the logic and motivation here

File: src/Compilers/CSharp/Portable/FlowAnalysis/PreciseAbstractFlowPass.cs [hunk 2]
```diff
@@ -1661,13 +1662,17 @@ private LocalState VisitSwitchHeader(BoundSwitchStatement node)
 
         private void VisitSwitchBlock(BoundSwitchStatement node)
         {
+            var afterSwitchState = UnreachableState();
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/FlowAnalysis/PreciseAbstractFlowPass.cs]: Consider adding a comment why UnreachableState is the right state to use here.

**Comment by AlekseyTs** [comment]
LGTM

File: src/Compilers/CSharp/Portable/FlowAnalysis/PreciseAbstractFlowPass.cs (hunk 2 above)
Comment (gafter) [inline src/Compilers/CSharp/Portable/FlowAnalysis/PreciseAbstractFlowPass.cs]: That would be a bit out of place in this source file (nearly every line could have such a comment). Every statement is unreachable unless there is some way of reaching it.

File: src/Compilers/CSharp/Portable/FlowAnalysis/PreciseAbstractFlowPass.cs (hunk 1 above)
Comment (gafter) [inline src/Compilers/CSharp/Portable/FlowAnalysis/PreciseAbstractFlowPass.cs]: Will do.

**Comment by gafter** [comment]
@jaredpar @VSadov @agocke @VladimirReshetnikov Can I have one more review please?

**Comment by jaredpar** [comment]
:+1:

**Pull request merged**